### DISCUSSION TITLE
No restriction to private content via socket via public user

### DISCUSSION THREAD

**Discussion by litehacker** [strapi/strapi#7920] [Legacy Topics]
**About the bug**
This is a security issue or just I do something wrong. 
The project is about questions, which need to be sent by socket to a user if the user has Authenticated JWT. 
Yet I have not implemented any authentication to the socket, but whenever the socket access my private documents in
`question` (the private questions), Strapi does not give any warnings or errors for the user which should not see them, neither warns me anyhow.

**Steps to reproduce the behavior**

1. Create Question collection
2. According to this [update](https://github.com/strapi/strapi/issues/5869#issuecomment-664803469) for `socket.io`, I get a question and try to send to the user. `bootstrap.js`:
`socket.emit('hello', {message: await strapi.services.question.findOne()});`

**Expected behavior**
It is expected to get an error for trying to access the private collection without the right authentication.

**Screenshots**
<img width="665" alt="Screen Shot 2020-09-16 at 15 10 59" src="https://user-images.githubusercontent.com/17227327/93335341-591a3180-f837-11ea-8fbf-cf773b0c866b.png">
<img width="1049" alt="Screen Shot 2020-09-16 at 16 13 22" src="https://user-images.githubusercontent.com/17227327/93335527-a1395400-f837-11ea-9ac5-d02a0d086d62.png">
<img width="1056" alt="Screen Shot 2020-09-16 at 16 14 41" src="https://user-images.githubusercontent.com/17227327/93335598-c4640380-f837-11ea-8667-20aae4aaca4b.png">

<img width="886" alt="Screen Shot 2020-09-16 at 16 19 36" src="https://user-images.githubusercontent.com/17227327/93336272-98954d80-f838-11ea-8273-02c07019139e.png">

**Code snippets**
Frontend code React:
```javascript
import React, { useState, useEffect } from "react";
import socketIOClient from "socket.io-client";
const ENDPOINT = "http://localhost:1337";

function App() {
  const [response, setResponse] = useState();

  useEffect(() => {
    const socket = socketIOClient(ENDPOINT);
    socket.on('hello', (msg, cb) => {
      console.log('SOCKET HI')
      console.log(msg.message)
      setResponse('sinav icerigi: ' + JSON.stringify(msg.message))
    });
  }, []);

  return (
    <>
      <p>
        Response is:
      </p>
      <p>
        {response}
      </p>
    </>
  );
}

export default App;
```
Bootstrap.js in Strapi:
```javascript
'use strict';
module.exports = async () => {
  process.nextTick(() =>{
    var io = require('socket.io')(strapi.server);
    io.on('connection', async function(socket) {

      console.log(`a user connected`)
      // HERE IS THE VULNERABLE DATA
      socket.emit('hello', {message: await strapi.services.question.findOne()});

      // listen for user diconnect
      socket.on('disconnect', () =>{
        console.log('a user disconnected')
      });
    });
    strapi.io = io; // register socket io inside strapi main object to use it globally anywhere
  })
};
```
**System**

- Node.js version: v12
- NPM version: 6.14.8
- Strapi version: 3.1.4
- Database: sqlite
- Operating system: mac os

**Comment by derrickmehaffy** [comment]
Hello,

I see you are wanting to ask a question that is not really a bug report or a feature request, questions should be directed to the [discussions](https://github.com/strapi/strapi/discussions). Please see the following contributing guidelines for asking a question [here](https://github.com/strapi/strapi/blob/master/CONTRIBUTING.md#reporting-an-issue).

I have converted this to a discussion for you, we try to keep Github clean and only for bug reports and feature requests. In your case it's a custom modification that doesn't ship with Strapi by default.

**Comment by x0ptr** [comment]
You're just running socket.io in bootstrap.js there is no authentication between client and server there.

Services are not handling authentication by itself. 
- https://strapi.io/documentation/v3.x/concepts/services.html#concept

An easy way to authenticate over sockets would be nice. But you dont need a warning if you just calls your services.

**Update**

example how you can verify the user 

```js
module.exports = () => {
  process.nextTick(() => {
    var io = require("socket.io")(strapi.server);
    io.on("connection", function (socket) {

      // add user
      socket.on("verify", (data) => {
        strapi.plugins["users-permissions"].services.jwt
          .verify(data)
          .then((data) => {
            if (data.id) {
              socket.join("basic");
            }
          })
          .catch((error) => {
            console.log(error);
          });
      });

      // remove user
      socket.on("disconnect", () => {
        socket.leave('basic');
      });
    });
    strapi.io = io;
  });
};
```

it only checks if the user is a authenticated user. Basicly its just verify the jwt. But here you could also check the roles etc. like this.
Each role could be in another room. That should work.

So if u want to message some events with you can emit messages from the controller to user. Because strapi.io is global.
